core/src/main/scala/akka/persistence/r2dbc/internal/h2/H2Dialect.scala

```diff
@@ -0,0 +1,165 @@
+      system: ActorSystem[_]): DurableStateDao =
+    new H2DurableStateDao(settings, connectionFactory)(ecForDaos(system), system)
+
+  private def ecForDaos(system: ActorSystem[_]): ExecutionContext = {
+    // H2 R2DBC driver blocks in surprising places (Mono.toFuture in stmt.execute().asFuture())
+    // FIXME is the default blocking good enough, should we have a separate to compartmentalize?
+    system.dispatchers.lookup(DispatcherSelector.blocking())
+  }
+
+  private def dbSchema(settings: R2dbcSettings, createSliceIndexes: Boolean): String = {
+    val sliceIndexes = if (createSliceIndexes) {
+      val sliceIndexWithSchema = settings.journalTableWithSchema + "_slice_idx"
+      val durableStateSliceIndexWithSchema = settings.durableStateTableWithSchema + "_slice_idx"
+      Seq(
+        sql"""
+             CREATE INDEX IF NOT EXISTS $sliceIndexWithSchema ON ${settings.journalTableWithSchema}(slice, entity_type, db_timestamp, seq_nr)""",
+        sql"""CREATE INDEX IF NOT EXISTS $durableStateSliceIndexWithSchema ON durable_state(slice, entity_type, db_timestamp, revision)""")
+    } else Seq.empty[String]
+
+    (Seq(
+      sql"""CREATE TABLE IF NOT EXISTS ${settings.journalTableWithSchema}(
+        slice INT NOT NULL,
+        entity_type VARCHAR(255) NOT NULL,
+        persistence_id VARCHAR(255) NOT NULL,
+        seq_nr BIGINT NOT NULL,
+        db_timestamp timestamp with time zone NOT NULL,
+
+        event_ser_id INTEGER NOT NULL,
+        event_ser_manifest VARCHAR(255) NOT NULL,
+        event_payload BYTEA NOT NULL,
+
+        deleted BOOLEAN DEFAULT FALSE NOT NULL,
+        writer VARCHAR(255) NOT NULL,
+        adapter_manifest VARCHAR(255),
+        tags TEXT ARRAY,
+
+        meta_ser_id INTEGER,
+        meta_ser_manifest VARCHAR(255),
+        meta_payload BYTEA,
+
+        PRIMARY KEY(persistence_id, seq_nr)
+      )""",
+      sql"""
+        CREATE TABLE IF NOT EXISTS ${settings.snapshotsTableWithSchema}(
+          slice INT NOT NULL,
+          entity_type VARCHAR(255) NOT NULL,
+          persistence_id VARCHAR(255) NOT NULL,
+          seq_nr BIGINT NOT NULL,
+          write_timestamp BIGINT NOT NULL,
+          ser_id INTEGER NOT NULL,
+          ser_manifest VARCHAR(255) NOT NULL,
+          snapshot BYTEA NOT NULL,
+          meta_ser_id INTEGER,
+          meta_ser_manifest VARCHAR(255),
+          meta_payload BYTEA,
+
+          PRIMARY KEY(persistence_id)
+        )""",
+      sql"""
+        CREATE TABLE IF NOT EXISTS ${settings.durableStateTableWithSchema} (
+          slice INT NOT NULL,
+          entity_type VARCHAR(255) NOT NULL,
+          persistence_id VARCHAR(255) NOT NULL,
+          revision BIGINT NOT NULL,
+          db_timestamp timestamp with time zone NOT NULL,
+
+          state_ser_id INTEGER NOT NULL,
+          state_ser_manifest VARCHAR(255),
+          state_payload BYTEA NOT NULL,
+          tags TEXT ARRAY,
+
```

Comment:
I introduced an `additional-init` setting for such things yesterday

Reply:
can we define all tables in config? is it possible to have such "huge" config block without escaping mess? Maybe the config can point to a resource file (in classpath) that has the table definitions and that is loaded?

Reply:
Actually, since the additional-init can alter tables it could add columns to the default table as well.

The table names need to come from other config-keys so tried to work with pulling those names into the h2 config instead, not entirely sure that will work out.

Reply:
alter table solves the problem with addition columns, indeed

Reply:
I need to try that out and see that there are no surprises there, and document it, though.